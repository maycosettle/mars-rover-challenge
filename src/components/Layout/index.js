export * from './PageLayout';
export * from './PaperLayout';
export * from './PhotoLayout';
