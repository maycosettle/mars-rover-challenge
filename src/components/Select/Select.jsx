import PropTypes from 'prop-types';
import { Fragment } from 'react';
import { Listbox, Transition } from '@headlessui/react';
import { CheckIcon, SelectorIcon } from '@heroicons/react/solid';

import useStyles from './styles';

const Select = ({ label, placeholder, disabled, options, selected, setSelected }) => {
  const styles = useStyles({ disabled });

  return (
    <Listbox value={selected} onChange={setSelected} disabled={disabled}>
      <Listbox.Label className={styles.label}>{label}</Listbox.Label>
      <div className="mt-1 relative">
        <Listbox.Button className={styles.button}>
          <span className={styles.name}>{selected.name || placeholder}</span>
          <span className={styles.iconContainer}>
            <SelectorIcon className={styles.icon} aria-hidden="true" />
          </span>
        </Listbox.Button>

        <Transition as={Fragment} leave="transition ease-in duration-100" leaveFrom="opacity-100" leaveTo="opacity-0">
          <Listbox.Options className={styles.optionsContainer}>
            {options.map((option, index) => (
              <Listbox.Option key={index.toString()} className={({ active }) => styles.active(active)} value={option}>
                {({ selected, active }) => (
                  <>
                    <span className={styles.optionName}>{option.name}</span>

                    {selected ? (
                      <span className={styles.selectedIconContainer(active)}>
                        <CheckIcon className={styles.selectedIcon} aria-hidden="true" />
                      </span>
                    ) : null}
                  </>
                )}
              </Listbox.Option>
            ))}
          </Listbox.Options>
        </Transition>
      </div>
    </Listbox>
  );
};

Select.propTypes = {
  label: PropTypes.string,
};

Select.defaultProps = {
  label: '',
};

export default Select;
