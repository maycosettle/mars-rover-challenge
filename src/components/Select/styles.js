import classNames from 'classnames';

const useStyles = ({ disabled }) => ({
  label: 'block text-sm font-medium text-gray-700',
  button: classNames(
    'cursor-pointer relative w-full bg-white border border-gray-300 rounded-md shadow-sm pl-3 pr-10 py-2 text-left focus:outline-none focus:ring-1 focus:ring-primary-500 focus:border-primary-500 sm:text-sm',
    {
      'cursor-not-allowed': disabled,
    }
  ),
  name: 'block truncate h-5',
  iconContainer: 'absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none',
  icon: 'h-5 w-5 text-gray-400',
  optionsContainer:
    'absolute z-10 mt-1 w-full bg-white shadow-lg max-h-60 rounded-md py-1 text-base ring-1 ring-black ring-opacity-5 overflow-auto focus:outline-none sm:text-sm',
  active: (active) =>
    classNames('cursor-pointer select-none relative py-2 pl-8 pr-4', {
      'text-white bg-primary-600': active,
      'text-gray-900': !active,
    }),
  optionName: (selected) =>
    classNames('block truncate font-normal', {
      'font-semibold': selected,
    }),
  selectedIconContainer: (active) =>
    classNames('absolute inset-y-0 right-0 flex items-center pr-4', {
      'text-white': active,
      'text-primary-600': !active,
    }),
  selectedIcon: 'h-5 w-5',
});

export default useStyles;
