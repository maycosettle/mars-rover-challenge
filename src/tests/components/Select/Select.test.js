import { shallow } from 'enzyme';
import { Select } from 'components/Select';

describe('tests on <Select />', () => {
  const options = [
    { value: 'FHAZ', name: 'FHAZ', fullName: 'Front Hazard Avoidance Camera' },
    { value: 'RHAZ', name: 'RHAZ', fullName: 'Rear Hazard Avoidance Camera' },
    { value: 'MAST', name: 'MAST', fullName: 'Mast Camera' },
  ];
  const wrapper = shallow(<Select options={options} selected={options[0]} onChange={jest.fn()} placeholder="test" />);

  test('Should render Select correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
