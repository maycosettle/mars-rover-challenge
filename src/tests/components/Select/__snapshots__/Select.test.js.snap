// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on <Select /> Should render Select correctly 1`] = `
<Listbox
  value={
    Object {
      "fullName": "Front Hazard Avoidance Camera",
      "name": "FHAZ",
      "value": "FHAZ",
    }
  }
>
  <Label
    className="block text-sm font-medium text-gray-700"
  />
  <div
    className="mt-1 relative"
  >
    <Button
      className="cursor-pointer relative w-full bg-white border border-gray-300 rounded-md shadow-sm pl-3 pr-10 py-2 text-left focus:outline-none focus:ring-1 focus:ring-primary-500 focus:border-primary-500 sm:text-sm"
    >
      <span
        className="block truncate h-5"
      >
        FHAZ
      </span>
      <span
        className="absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none"
      >
        <SelectorIcon
          aria-hidden="true"
          className="h-5 w-5 text-gray-400"
        />
      </span>
    </Button>
    <Transition
      as={Symbol(react.fragment)}
      leave="transition ease-in duration-100"
      leaveFrom="opacity-100"
      leaveTo="opacity-0"
    >
      <Options
        className="absolute z-10 mt-1 w-full bg-white shadow-lg max-h-60 rounded-md py-1 text-base ring-1 ring-black ring-opacity-5 overflow-auto focus:outline-none sm:text-sm"
      >
        <Option
          className={[Function]}
          key="0"
          value={
            Object {
              "fullName": "Front Hazard Avoidance Camera",
              "name": "FHAZ",
              "value": "FHAZ",
            }
          }
        >
          <Component />
        </Option>
        <Option
          className={[Function]}
          key="1"
          value={
            Object {
              "fullName": "Rear Hazard Avoidance Camera",
              "name": "RHAZ",
              "value": "RHAZ",
            }
          }
        >
          <Component />
        </Option>
        <Option
          className={[Function]}
          key="2"
          value={
            Object {
              "fullName": "Mast Camera",
              "name": "MAST",
              "value": "MAST",
            }
          }
        >
          <Component />
        </Option>
      </Options>
    </Transition>
  </div>
</Listbox>
`;
